### ISSUE TITLE
Excessive disk space being used by `wallet_block_db`

### ISSUE THREAD

**Issue by Sword-Smith** [Neptune-Crypto/neptune-core#14]
After ~1900 blocks my node ran out of disk space. I only had 1.8 GB available, but all that was used by the `~/.local/share/neptune/alpha/databases/wallet_block_db` directory. 

The archival mutator set was only 1.7MB, and all blocks stored on disk was only 33MB.

Investigate why `wallet_block_db` is using so much disk space, and fix the issue.

**Comment by Sword-Smith** [comment]
Now the same directory is 11 MB without doing anything other than opening and closing the wallet database a few times. Seems like `rusty-leveldb` was able to clean up unused files. The issue seems to be with `rusty-leveldb`. Maybe there is a command we can run that will prune unused files?

**Comment by Sword-Smith** [comment]
The relevant code in `persist` interacts with the underlying `rusty-leveldb` database through this code

```rust
fn persist(&mut self) {
        let mut write_batch = WriteBatch::new();

        for table in &self.schema.tables {
            let operations = table
                .lock()
                .expect("Could not obtain lock on table in RustyWallet as StorageWriter (persist).")
                .pull_queue();
            for op in operations {
                match op {
                    WriteOperation::Write(key, value) => write_batch.put(&key.0, &value.0),
                    WriteOperation::Delete(key) => write_batch.delete(&key.0),
                }
            }
        }

        self.db
            .lock()
            .expect("RustyWallet: StorageWriter -- could not get lock on database for writing.")
            .write(write_batch, true)
            .expect("Could not persist to database.");
```

I would expect the call to `write(write_batch, true)` to eventually run `delete_obsolete_files`, but that doesn't seem to be the case. Is an extra call to `flush` necessary? Attemting that solution in the branch `reduce-wallet-disk-usage-issue-14`.

**Comment by Sword-Smith** [comment]
Adding a `flush` of the database seems to fix the problem. I don't understand why though. Maybe @dermesser can explain why?

```rust
 fn persist(&mut self) {
        let mut write_batch = WriteBatch::new();

        for table in &self.schema.tables {
            let operations = table
                .lock()
                .expect("Could not obtain lock on table in RustyWallet as StorageWriter (persist).")
                .pull_queue();
            for op in operations {
                match op {
                    WriteOperation::Write(key, value) => write_batch.put(&key.0, &value.0),
                    WriteOperation::Delete(key) => write_batch.delete(&key.0),
                }
            }
        }

        self.db
            .lock()
            .expect("RustyWallet: StorageWriter -- could not get lock on database for writing.")
            .write(write_batch, true)
            .expect("Could not persist to database.");

        // Call `flush` to enforce compaction to reduce disk space
        self.db
            .lock()
            .expect("Must be able to acquire DB lock in persist")
            .flush()
            .expect("Flushing must succeed in persist");
    }
```

**Comment by dermesser** [comment]
Yes, in principle writing should eventually trigger a compaction. Of course that's hard to reproduce and may depend on the specific usage pattern. A very large memtable (`write_buffer_size` option) would delay rotating the log though, but once a memtable has been written to an L0 table, the log should be deleted. Maybe you can verify this at some point by looking at the file system.

Flushing is a good idea anyway, if you rely on persistence. When writing several records at once, the performance penalty might not be prohibitive.

**Comment by Sword-Smith** [comment]
> Flushing is a good idea anyway, if you rely on persistence. When writing several records at once, the performance penalty might not be prohibitive.
But shouldn't the call to `.write(write_batch, true)` trigger the `flush`. Will it make a difference if I add a subsequent call to `flush`?

**Comment by Sword-Smith** [comment]
See to be resolved by the merge of #16. Will leave this issue open till others confirm.

**Comment by Sword-Smith** [comment]
We're back at ~700MB now, unfortunately

```bash
$date; ls -lah /home/thv/.local/share/neptune/alpha/databases/wallet_block_db/
Wed 12 Jul 2023 04:30:34 PM UTC
total 688M                                       
drwxr-xr-x 2 thv thv  12K Jul 12 16:29 .         
drwxr-xr-x 6 thv thv 4.0K Jul  7 12:03 ..        
-rw-r--r-- 1 thv thv 2.1M Jul 12 14:40 003736.ldb
-rw-r--r-- 1 thv thv 2.1M Jul 12 14:44 003748.ldb
-rw-r--r-- 1 thv thv 2.1M Jul 12 14:49 003760.ldb
-rw-r--r-- 1 thv thv 807K Jul 12 14:51 003772.ldb
-rw-r--r-- 1 thv thv 2.1M Jul 12 14:51 003773.ldb
-rw-r--r-- 1 thv thv 2.1M Jul 12 14:58 003784.ldb
-rw-r--r-- 1 thv thv 2.1M Jul 12 15:02 003803.ldb     
-rw-r--r-- 1 thv thv 2.1M Jul 12 15:02 003804.ldb
-rw-r--r-- 1 thv thv 2.1M Jul 12 15:02 003803.ldb
-rw-r--r-- 1 thv thv 2.1M Jul 12 15:02 003804.ldb
-rw-r--r-- 1 thv thv 2.1M Jul 12 15:02 003805.ldb
-rw-r--r-- 1 thv thv 2.1M Jul 12 15:02 003806.ldb
-rw-r--r-- 1 thv thv 2.1M Jul 12 15:02 003807.ldb
-rw-r--r-- 1 thv thv 808K Jul 12 15:02 003808.ldb
-rw-r--r-- 1 thv thv  11M Jul 12 15:02 003810.ldb
-rw-r--r-- 1 thv thv  11M Jul 12 15:04 003812.ldb
-rw-r--r-- 1 thv thv  11M Jul 12 15:05 003814.ldb
-rw-r--r-- 1 thv thv  11M Jul 12 15:06 003816.ldb
-rw-r--r-- 1 thv thv  11M Jul 12 15:08 003818.ldb
-rw-r--r-- 1 thv thv  11M Jul 12 15:10 003820.ldb
-rw-r--r-- 1 thv thv  11M Jul 12 15:15 003822.ldb
-rw-r--r-- 1 thv thv  11M Jul 12 15:16 003824.ldb
-rw-r--r-- 1 thv thv  11M Jul 12 15:17 003826.ldb
-rw-r--r-- 1 thv thv  11M Jul 12 15:20 003828.ldb
-rw-r--r-- 1 thv thv  11M Jul 12 15:21 003830.ldb
-rw-r--r-- 1 thv thv  11M Jul 12 15:21 003832.ldb
-rw-r--r-- 1 thv thv  11M Jul 12 15:26 003834.ldb
-rw-r--r-- 1 thv thv  11M Jul 12 15:31 003836.ldb
-rw-r--r-- 1 thv thv  11M Jul 12 15:32 003838.ldb
-rw-r--r-- 1 thv thv  11M Jul 12 15:32 003840.ldb
-rw-r--r-- 1 thv thv  11M Jul 12 15:33 003842.ldb
-rw-r--r-- 1 thv thv  11M Jul 12 15:33 003844.ldb
-rw-r--r-- 1 thv thv  11M Jul 12 15:34 003846.ldb
-rw-r--r-- 1 thv thv  11M Jul 12 15:36 003848.ldb
-rw-r--r-- 1 thv thv  11M Jul 12 15:36 003850.ldb
-rw-r--r-- 1 thv thv  11M Jul 12 15:36 003852.ldb
-rw-r--r-- 1 thv thv  11M Jul 12 15:40 003854.ldb
-rw-r--r-- 1 thv thv  11M Jul 12 15:40 003856.ldb
-rw-r--r-- 1 thv thv  11M Jul 12 15:41 003858.ldb
-rw-r--r-- 1 thv thv  11M Jul 12 15:42 003860.ldb
-rw-r--r-- 1 thv thv  11M Jul 12 15:47 003862.ldb
-rw-r--r-- 1 thv thv  11M Jul 12 15:47 003864.ldb
-rw-r--r-- 1 thv thv  11M Jul 12 15:49 003866.ldb
-rw-r--r-- 1 thv thv  11M Jul 12 15:50 003868.ldb
-rw-r--r-- 1 thv thv  11M Jul 12 15:51 003870.ldb
-rw-r--r-- 1 thv thv  11M Jul 12 15:53 003872.ldb
-rw-r--r-- 1 thv thv  11M Jul 12 15:54 003874.ldb
-rw-r--r-- 1 thv thv  11M Jul 12 15:56 003876.ldb
-rw-r--r-- 1 thv thv  11M Jul 12 15:59 003878.ldb
-rw-r--r-- 1 thv thv  11M Jul 12 15:59 003880.ldb
-rw-r--r-- 1 thv thv  11M Jul 12 15:59 003882.ldb
-rw-r--r-- 1 thv thv  11M Jul 12 16:00 003884.ldb
-rw-r--r-- 1 thv thv  11M Jul 12 16:03 003886.ldb
-rw-r--r-- 1 thv thv  11M Jul 12 16:05 003888.ldb
-rw-r--r-- 1 thv thv  11M Jul 12 16:06 003890.ldb
-rw-r--r-- 1 thv thv  11M Jul 12 16:07 003892.ldb
-rw-r--r-- 1 thv thv  11M Jul 12 16:10 003894.ldb
-rw-r--r-- 1 thv thv  11M Jul 12 16:11 003896.ldb
-rw-r--r-- 1 thv thv  11M Jul 12 16:12 003898.ldb
-rw-r--r-- 1 thv thv  11M Jul 12 16:12 003900.ldb
-rw-r--r-- 1 thv thv  11M Jul 12 16:12 003902.ldb
-rw-r--r-- 1 thv thv  11M Jul 12 16:16 003904.ldb
-rw-r--r-- 1 thv thv  11M Jul 12 16:17 003906.ldb
-rw-r--r-- 1 thv thv  11M Jul 12 16:17 003908.ldb
-rw-r--r-- 1 thv thv  11M Jul 12 16:20 003910.ldb
-rw-r--r-- 1 thv thv  11M Jul 12 16:22 003912.ldb
-rw-r--r-- 1 thv thv  11M Jul 12 16:22 003914.ldb
-rw-r--r-- 1 thv thv  11M Jul 12 16:25 003916.ldb
-rw-r--r-- 1 thv thv  11M Jul 12 16:26 003918.ldb
-rw-r--r-- 1 thv thv  11M Jul 12 16:26 003920.ldb
-rw-r--r-- 1 thv thv  11M Jul 12 16:26 003922.ldb
-rw-r--r-- 1 thv thv  11M Jul 12 16:27 003924.ldb
-rw-r--r-- 1 thv thv  11M Jul 12 16:27 003926.ldb
-rw-r--r-- 1 thv thv  11M Jul 12 16:29 003928.ldb
-rw-r--r-- 1 thv thv   38 Jul 12 16:30 003929.log
-rw-r--r-- 1 thv thv  11M Jul 12 16:29 003930.ldb
-rw-r--r-- 1 thv thv   16 Jul 12 16:29 CURRENT
-rw-r--r-- 1 thv thv    0 Jul  7 12:03 LOCK
-rw-r--r-- 1 thv thv 107K Jul 12 16:29 LOG
-rw-r--r-- 1 thv thv 9.8K Jul 12 09:46 LOG.old
-rw-r--r-- 1 thv thv 148K Jul 12 16:29 MANIFEST-000001
```

**Comment by Sword-Smith** [comment]
After forcing a cleanup by stopping the node and starting it again, the directory looks like this

```bash
$ date; ls -lah /home/thv/.local/share/neptune/alpha/databases/wallet_block_db/
Wed 12 Jul 2023 04:44:13 PM UTC
total 44M
drwxr-xr-x 2 thv thv  12K Jul 12 16:44 .
drwxr-xr-x 6 thv thv 4.0K Jul  7 12:03 ..
-rw-r--r-- 1 thv thv 2.1M Jul 12 14:40 003736.ldb
-rw-r--r-- 1 thv thv 2.1M Jul 12 14:44 003748.ldb
-rw-r--r-- 1 thv thv 2.1M Jul 12 14:49 003760.ldb
-rw-r--r-- 1 thv thv 807K Jul 12 14:51 003772.ldb
-rw-r--r-- 1 thv thv 2.1M Jul 12 14:51 003773.ldb
-rw-r--r-- 1 thv thv 2.1M Jul 12 14:58 003784.ldb
-rw-r--r-- 1 thv thv 2.1M Jul 12 16:43 003954.ldb
-rw-r--r-- 1 thv thv 2.1M Jul 12 16:43 003955.ldb
-rw-r--r-- 1 thv thv 2.1M Jul 12 16:44 003956.ldb
-rw-r--r-- 1 thv thv 2.1M Jul 12 16:44 003957.ldb
-rw-r--r-- 1 thv thv 2.1M Jul 12 16:44 003958.ldb
-rw-r--r-- 1 thv thv 921K Jul 12 16:44 003959.ldb
-rw-r--r-- 1 thv thv  11M Jul 12 16:44 003961.ldb
-rw-r--r-- 1 thv thv   19 Jul 12 16:44 003962.log
-rw-r--r-- 1 thv thv  11M Jul 12 16:44 003963.ldb
-rw-r--r-- 1 thv thv   16 Jul 12 16:44 CURRENT
-rw-r--r-- 1 thv thv    0 Jul  7 12:03 LOCK
-rw-r--r-- 1 thv thv 4.1K Jul 12 16:44 LOG
-rw-r--r-- 1 thv thv  640 Jul 12 16:43 LOG.old
-rw-r--r-- 1 thv thv 150K Jul 12 16:44 MANIFEST-000001

```

**Comment by Sword-Smith** [comment]
Output from my running client the last 12 hours
```bash
while true; do date; neptune-cli block-height; du -h /home/thv/.local/share/neptune/alpha/databases/wallet_block_db/; sleep 900; done                                                                                                        
Wed 12 Jul 2023 05:05:51 PM UTC                                     
Block height: 00000000000000002523                                  
132M    /home/thv/.local/share/neptune/alpha/databases/wallet_block_db/                                                                  
Wed 12 Jul 2023 05:20:51 PM UTC                                                                                                          
Block height: 00000000000000002535                                                                                                       
264M    /home/thv/.local/share/neptune/alpha/databases/wallet_block_db/                                                                                                                                                                                                           
Wed 12 Jul 2023 05:35:51 PM UTC                                                                                                          
Block height: 00000000000000002547                                                                                                       
397M    /home/thv/.local/share/neptune/alpha/databases/wallet_block_db/                                                                                                                                                                                                           
Wed 12 Jul 2023 05:50:51 PM UTC                                                                                                          
Block height: 00000000000000002554                                                                                                       
477M    /home/thv/.local/share/neptune/alpha/databases/wallet_block_db/                                                                                                                                                                                                           
Wed 12 Jul 2023 06:05:51 PM UTC                                                                                                          
Block height: 00000000000000002554                                                                                                       
477M    /home/thv/.local/share/neptune/alpha/databases/wallet_block_db/                                                                                                                                                                                                           
Wed 12 Jul 2023 06:20:51 PM UTC                                                                                                          
Block height: 00000000000000002554                                                                                                       
477M    /home/thv/.local/share/neptune/alpha/databases/wallet_block_db/                                                                                                                                                                                                           
Wed 12 Jul 2023 06:35:51 PM UTC                                                                                                          
Block height: 00000000000000002554                                                                                                       
477M    /home/thv/.local/share/neptune/alpha/databases/wallet_block_db/                                                                                                                                                                                                           
Wed 12 Jul 2023 06:50:51 PM UTC                                                                                                          
Block height: 00000000000000002554                                                                                                       
477M    /home/thv/.local/share/neptune/alpha/databases/wallet_block_db/                                                                                                                                                                                                           
Wed 12 Jul 2023 07:05:51 PM UTC                                                                                                          
Block height: 00000000000000002556                                                                                                       
499M    /home/thv/.local/share/neptune/alpha/databases/wallet_block_db/                                                                                                                                                                                                           
Wed 12 Jul 2023 07:20:51 PM UTC                                     
Block height: 00000000000000002556                                  
499M    /home/thv/.local/share/neptune/alpha/databases/wallet_block_db/                                                                  
Wed 12 Jul 2023 07:35:51 PM UTC                                     
Block height: 00000000000000002556                                  
499M    /home/thv/.local/share/neptune/alpha/databases/wallet_block_db/                                                                  
Wed 12 Jul 2023 07:50:51 PM UTC                                     
Block height: 00000000000000002556                                  
499M    /home/thv/.local/share/neptune/alpha/databases/wallet_block_db/                                                                  
Wed 12 Jul 2023 08:05:51 PM UTC                                     
Block height: 00000000000000002557                                  
511M    /home/thv/.local/share/neptune/alpha/databases/wallet_block_db/                                                                  
Wed 12 Jul 2023 08:20:51 PM UTC                                     
Block height: 00000000000000002557                                  
511M    /home/thv/.local/share/neptune/alpha/databases/wallet_block_db/                                                                  
Wed 12 Jul 2023 08:35:51 PM UTC                                     
Block height: 00000000000000002557                                  
511M    /home/thv/.local/share/neptune/alpha/databases/wallet_block_db/                                                                  
Wed 12 Jul 2023 08:50:51 PM UTC                                     
Block height: 00000000000000002557                                  
511M    /home/thv/.local/share/neptune/alpha/databases/wallet_block_db/                                                                  
Wed 12 Jul 2023 09:05:51 PM UTC                                     
Block height: 00000000000000002559                                  
533M    /home/thv/.local/share/neptune/alpha/databases/wallet_block_db/                                                                  
Wed 12 Jul 2023 09:20:51 PM UTC                                     
Block height: 00000000000000002569                                  
647M    /home/thv/.local/share/neptune/alpha/databases/wallet_block_db/                                                                  
Wed 12 Jul 2023 09:35:51 PM UTC
Block height: 00000000000000002576                                  
726M    /home/thv/.local/share/neptune/alpha/databases/wallet_block_db/                                                                  
Wed 12 Jul 2023 09:50:51 PM UTC                                     
Block height: 00000000000000002587                                  
851M    /home/thv/.local/share/neptune/alpha/databases/wallet_block_db/                                                                  
Wed 12 Jul 2023 10:05:51 PM UTC                                     
Block height: 00000000000000002596                                  
953M    /home/thv/.local/share/neptune/alpha/databases/wallet_block_db/                                                                  
Wed 12 Jul 2023 10:20:51 PM UTC                                     
Block height: 00000000000000002603                                  
1.1G    /home/thv/.local/share/neptune/alpha/databases/wallet_block_db/                                                                  
Wed 12 Jul 2023 10:35:51 PM UTC                                     
Block height: 00000000000000002603                                  
1.1G    /home/thv/.local/share/neptune/alpha/databases/wallet_block_db/                                                                  
Wed 12 Jul 2023 10:50:51 PM UTC                                     
Block height: 00000000000000002603                                  
1.1G    /home/thv/.local/share/neptune/alpha/databases/wallet_block_db/                                                                                                                                                                                                           
Wed 12 Jul 2023 11:05:51 PM UTC                                     
Block height: 00000000000000002604                                  
1.1G    /home/thv/.local/share/neptune/alpha/databases/wallet_block_db/                                                                  
Wed 12 Jul 2023 11:20:51 PM UTC                                                                                                          
Block height: 00000000000000002604                                                                                                       
1.1G    /home/thv/.local/share/neptune/alpha/databases/wallet_block_db/                                                                                                                                                                                                           
Wed 12 Jul 2023 11:35:51 PM UTC                                                                                                          
Block height: 00000000000000002604                                                                                                       
1.1G    /home/thv/.local/share/neptune/alpha/databases/wallet_block_db/                                                                                                                                                                                                           
Wed 12 Jul 2023 11:50:51 PM UTC                                                                                                          
Block height: 00000000000000002604                                                                                                       
1.1G    /home/thv/.local/share/neptune/alpha/databases/wallet_block_db/                                                                                                                                                                                                           
Thu 13 Jul 2023 12:05:51 AM UTC                                                                                                          
Block height: 00000000000000002604                                                                                                       
23M     /home/thv/.local/share/neptune/alpha/databases/wallet_block_db/                                                                                                                                                                                                           
Thu 13 Jul 2023 12:20:51 AM UTC                                                                                                          
Block height: 00000000000000002604                                                                                                       
23M     /home/thv/.local/share/neptune/alpha/databases/wallet_block_db/                                                                                                                                                                                                           
Thu 13 Jul 2023 12:35:51 AM UTC                                                                                                          
Block height: 00000000000000002605                                                                                                       
34M     /home/thv/.local/share/neptune/alpha/databases/wallet_block_db/                                                                                                                                                                                                           
Thu 13 Jul 2023 12:50:51 AM UTC                                                                                                          
Block height: 00000000000000002605                                                                                                       
34M     /home/thv/.local/share/neptune/alpha/databases/wallet_block_db/                                                                                                                                                                                                           
Thu 13 Jul 2023 01:05:51 AM UTC                                                                                                          
Block height: 00000000000000002605                                                                                                       
34M     /home/thv/.local/share/neptune/alpha/databases/wallet_block_db/                                                                                                                                                                                                           
Thu 13 Jul 2023 01:20:51 AM UTC                                     
Block height: 00000000000000002605                                  
34M     /home/thv/.local/share/neptune/alpha/databases/wallet_block_db/                                                                  
Thu 13 Jul 2023 01:35:51 AM UTC                                     
Block height: 00000000000000002605                                  
34M     /home/thv/.local/share/neptune/alpha/databases/wallet_block_db/                                                                  
Thu 13 Jul 2023 01:50:51 AM UTC                                     
Block height: 00000000000000002605                                  
34M     /home/thv/.local/share/neptune/alpha/databases/wallet_block_db/                                                                  
Thu 13 Jul 2023 02:05:51 AM UTC                                     
Block height: 00000000000000002605                                  
34M     /home/thv/.local/share/neptune/alpha/databases/wallet_block_db/                                                                  
Thu 13 Jul 2023 02:20:51 AM UTC                                     
Block height: 00000000000000002606                                  
46M     /home/thv/.local/share/neptune/alpha/databases/wallet_block_db/                                                                  
Thu 13 Jul 2023 02:35:51 AM UTC                                     
Block height: 00000000000000002606                                  
46M     /home/thv/.local/share/neptune/alpha/databases/wallet_block_db/                                                                  
Thu 13 Jul 2023 02:50:51 AM UTC                                     
Block height: 00000000000000002607                                  
57M     /home/thv/.local/share/neptune/alpha/databases/wallet_block_db/                                                                  
Thu 13 Jul 2023 03:05:51 AM UTC                                     
Block height: 00000000000000002608                                  
68M     /home/thv/.local/share/neptune/alpha/databases/wallet_block_db/                                                                  
Thu 13 Jul 2023 03:20:51 AM UTC
Block height: 00000000000000002608                                                                                                                                                                                                                                        [1/1008]
68M     /home/thv/.local/share/neptune/alpha/databases/wallet_block_db/                                                                  
Thu 13 Jul 2023 03:35:51 AM UTC                                     
Block height: 00000000000000002608                                  
68M     /home/thv/.local/share/neptune/alpha/databases/wallet_block_db/                                                                  
Thu 13 Jul 2023 03:50:51 AM UTC                                     
Block height: 00000000000000002608                                  
68M     /home/thv/.local/share/neptune/alpha/databases/wallet_block_db/                                                                  
Thu 13 Jul 2023 04:05:51 AM UTC                                     
Block height: 00000000000000002608                                  
68M     /home/thv/.local/share/neptune/alpha/databases/wallet_block_db/                                                                  
Thu 13 Jul 2023 04:20:51 AM UTC                                     
Block height: 00000000000000002608                                  
68M     /home/thv/.local/share/neptune/alpha/databases/wallet_block_db/                                                                  
Thu 13 Jul 2023 04:35:51 AM UTC                                     
Block height: 00000000000000002608                                  
68M     /home/thv/.local/share/neptune/alpha/databases/wallet_block_db/                                                                  
Thu 13 Jul 2023 04:50:51 AM UTC                                     
Block height: 00000000000000002608                                  
68M     /home/thv/.local/share/neptune/alpha/databases/wallet_block_db/                                                                  
Thu 13 Jul 2023 05:05:51 AM UTC                                     
Block height: 00000000000000002608                                  
68M     /home/thv/.local/share/neptune/alpha/databases/wallet_block_db/                                                                  
Thu 13 Jul 2023 05:20:51 AM UTC                                     
Block height: 00000000000000002608                                  
68M     /home/thv/.local/share/neptune/alpha/databases/wallet_block_db/                                                                  
Thu 13 Jul 2023 05:35:51 AM UTC                                     
Block height: 00000000000000002608                                  
68M     /home/thv/.local/share/neptune/alpha/databases/wallet_block_db/                                                                  
Thu 13 Jul 2023 05:50:51 AM UTC                                     
Block height: 00000000000000002609                                  
80M     /home/thv/.local/share/neptune/alpha/databases/wallet_block_db/                                                                  
Thu 13 Jul 2023 06:05:51 AM UTC                                     
Block height: 00000000000000002609                                  
80M     /home/thv/.local/share/neptune/alpha/databases/wallet_block_db/                                                                  
Thu 13 Jul 2023 06:20:51 AM UTC                                     
Block height: 00000000000000002610                                  
91M     /home/thv/.local/share/neptune/alpha/databases/wallet_block_db/                                                                  
Thu 13 Jul 2023 06:35:51 AM UTC                                     
Block height: 00000000000000002610                                  
91M     /home/thv/.local/share/neptune/alpha/databases/wallet_block_db/                                                                  
Thu 13 Jul 2023 06:50:51 AM UTC                                     
Block height: 00000000000000002610                                  
91M     /home/thv/.local/share/neptune/alpha/databases/wallet_block_db/                                                                  
Thu 13 Jul 2023 07:05:51 AM UTC                                     
Block height: 00000000000000002611                                  
103M    /home/thv/.local/share/neptune/alpha/databases/wallet_block_db/                                                                  
Thu 13 Jul 2023 07:20:51 AM UTC                                     
Block height: 00000000000000002611                                  
103M    /home/thv/.local/share/neptune/alpha/databases/wallet_block_db/                                                                  
Thu 13 Jul 2023 07:35:51 AM UTC                                     
Block height: 00000000000000002611                                  
103M    /home/thv/.local/share/neptune/alpha/databases/wallet_block_db/                                                                  
Thu 13 Jul 2023 07:50:51 AM UTC                                     
Block height: 00000000000000002612                                  
114M    /home/thv/.local/share/neptune/alpha/databases/wallet_block_db/                                                                  
Thu 13 Jul 2023 08:05:51 AM UTC                                     
Block height: 00000000000000002618                                  
182M    /home/thv/.local/share/neptune/alpha/databases/wallet_block_db/                                                                  
Thu 13 Jul 2023 08:20:51 AM UTC                                     
Block height: 00000000000000002625                                  
262M    /home/thv/.local/share/neptune/alpha/databases/wallet_block_db/                                                                  
Thu 13 Jul 2023 08:35:51 AM UTC                                     
Block height: 00000000000000002636                                  
387M    /home/thv/.local/share/neptune/alpha/databases/wallet_block_db/                                                                  
Thu 13 Jul 2023 08:50:51 AM UTC                                     
Block height: 00000000000000002644                                  
478M    /home/thv/.local/share/neptune/alpha/databases/wallet_block_db/                                                                  
Thu 13 Jul 2023 09:05:51 AM UTC                                     
Block height: 00000000000000002655                                  
604M    /home/thv/.local/share/neptune/alpha/databases/wallet_block_db/                                                                  
Thu 13 Jul 2023 09:20:51 AM UTC                                     
Block height: 00000000000000002656
Block height: 00000000000000002656
615M    /home/thv/.local/share/neptune/alpha/databases/wallet_block_db/
Thu 13 Jul 2023 09:35:51 AM UTC
Block height: 00000000000000002657
626M    /home/thv/.local/share/neptune/alpha/databases/wallet_block_db/
```

**Comment by dermesser** [comment]
The thing is, I'm also confused as to why `flush()` would make a difference. For all I've seen, the code is merely using a standard file wrapped inside a `BufWriter`, which definitely should not take up disk space if not flushed (memory yes, but disk??)

To some extent, what you are seeing appears natural (unfortunately), as old tables are not compacted as often as you hope. This also has upsides, for example it avoids noticeable pauses: compacting those old tables will take more than just a few milliseconds.

Due to the levels structure, older tables are compacted exponentially less frequently and thus accumulate old keys. I wonder if we could tune this, to favor more frequent (automatic) clean-ups. However, by its nature as a log-structured database, there will always be some amount of dragging along of old keys. Currently, I believe the behavior of compaction especially of older levels (as I think this is what you're seeing, as opposed to L0 files), is controlled from https://github.com/dermesser/leveldb-rs/blob/master/src/version_set.rs#L527 - this defines an exponentially rising size of each level. We can add an option for this, to enforce higher scores and earlier compactions. In theory this should keep younger levels smaller, but may also push down the problem to older levels.

By the way, you don't happen to be holding snapshots? Those will also stop keys from being removed.

**Comment by Sword-Smith** [comment]
> Due to the levels structure, older tables are compacted exponentially less frequently and thus accumulate old keys. I wonder if we could tune this, to favor more frequent (automatic) clean-ups.
That might explain our problem. Our privacy-preserving data structure mutator sets requires the user to maintain membership proofs indefinitely. These membership proofs live in the DB that uses a lot of diskspace.

> (...) compacting those old tables will take more than just a few milliseconds.
Milliseconds is not a problem in our use case. When we receive a block, using it to update state takes some time (milliseconds to seconds), and a block only comes on average every 10 minutes. For our usecase paying a price of seconds would be very well worth it, if it could prevent a database from growing from ~20MB to at least 2GB as is currently the case.

> By the way, you don't happen to be holding snapshots? Those will also stop keys from being removed.

We are not using the snapshot functionality anywhere, no.

**Comment by aszepieniec** [comment]
Possibly related:

I noticed that my node crashed. The last logged message was
```
thread 'main' panicked at 'Could not persist to database.: Status { code: IOError, err: "IOError: No space left on device (os error 28)" }', src/models/state/wallet/rusty_wallet_database.rs:93:14
```

I tried to determine which database is eating up all this space:
```
alan@copernicus:~/neptune-core$ du -sh ~/.local/share/neptune/alpha/databases/banned_ips/
24K     /home/alan/.local/share/neptune/alpha/databases/banned_ips/
alan@copernicus:~/neptune-core$ du -sh ~/.local/share/neptune/alpha/databases/block_index/
1.9M    /home/alan/.local/share/neptune/alpha/databases/block_index/
alan@copernicus:~/neptune-core$ du -sh ~/.local/share/neptune/alpha/databases/mutator_set/
3.0M    /home/alan/.local/share/neptune/alpha/databases/mutator_set/
alan@copernicus:~/neptune-core$ du -sh ~/.local/share/neptune/alpha/databases/wallet_block_db/
6.5G    /home/alan/.local/share/neptune/alpha/databases/wallet_block_db/
```

The node was running a version that predates 689db5cab8d673b3b11d930e03442ba2fbf6d0be which possibly fixes it. I will try deleting the offending files and running the updated software.

**Comment by Sword-Smith** [comment]
> Possibly related:
> 
> I noticed that my node crashed. The last logged message was
> ```
> thread 'main' panicked at 'Could not persist to database.: Status { code: IOError, err: "IOError: No space left on device (os error 28)" }', src/models/state/wallet/rusty_wallet_database.rs:93:14
> ```
> 
> I tried to determine which database is eating up all this space:
> ```
> alan@copernicus:~/neptune-core$ du -sh ~/.local/share/neptune/alpha/databases/banned_ips/
> 24K     /home/alan/.local/share/neptune/alpha/databases/banned_ips/
> alan@copernicus:~/neptune-core$ du -sh ~/.local/share/neptune/alpha/databases/block_index/
> 1.9M    /home/alan/.local/share/neptune/alpha/databases/block_index/
> alan@copernicus:~/neptune-core$ du -sh ~/.local/share/neptune/alpha/databases/mutator_set/
> 3.0M    /home/alan/.local/share/neptune/alpha/databases/mutator_set/
> alan@copernicus:~/neptune-core$ du -sh ~/.local/share/neptune/alpha/databases/wallet_block_db/
> 6.5G    /home/alan/.local/share/neptune/alpha/databases/wallet_block_db/
> ```
> 
> The node was running a version that predates 689db5cab8d673b3b11d930e03442ba2fbf6d0be which possibly fixes it. I will try deleting the offending files and running the updated software.

The issue you see has already been fixed by @dermesser but hasn't been put into a new version of leveldb yet. When it does, we will have to update our dependency chain starting from twenty-first.

**Comment by Sword-Smith** [comment]
The fix by @dermesser seems to work. I have found no indication that it doesn't. I did run into a new error involving the number of open files. But havent been able to reproduce it:
```
thread 'main' panicked at 'Could not persist to database.: Status { code: IOError, err: "open (write): IOError: Too many open files (os error 24): /home/thv/.local/share/neptune/alpha/databases/wallet_block_db/000001.dbtmp" }', src/models/state/wallet/rusty_wallet_database.rs:93:14
```

**Comment by Sword-Smith** [comment]
Fixed by upgrading to rusty-leveldb 2.0.0
